server/routes/ad.ts

```diff
@@ -59,42 +60,85 @@ type PutDetectorParams = {
 };
 
 export function registerADRoutes(apiRouter: Router, adService: AdService) {
+  // create detector
   apiRouter.post('/detectors', adService.putDetector);
-  apiRouter.put('/detectors/{detectorId}', adService.putDetector);
+  apiRouter.post('/detectors/{dataSourceId}', adService.putDetector);
+
+  // put detector
+  apiRouter.put('/detectors/{detectorId}', 
+    adService.putDetector);
+  apiRouter.put('/detectors/{detectorId}/{dataSourceId}', 
+    adService.putDetector);
+
   apiRouter.post('/detectors/_search', adService.searchDetector);
-  apiRouter.post('/detectors/results/_search/', adService.searchResults);
-  apiRouter.post('/detectors/results/_search', adService.searchResults);
-  apiRouter.post(
-    '/detectors/results/_search/{resultIndex}/{onlyQueryCustomResultIndex}',
-    adService.searchResults
-  );
-  apiRouter.get('/detectors/{detectorId}', adService.getDetector);
-  apiRouter.get('/detectors', adService.getDetectors);
+  
+
+  // post search anomaly results
+  apiRouter.post('/detectors/results/_search/', 
+    adService.searchResults);
+  apiRouter.post('/detectors/results/{dataSourceId}/_search', 
+    adService.searchResults);
+  apiRouter.post('/detectors/results/_search/{resultIndex}/{onlyQueryCustomResultIndex}',
+    adService.searchResults);
+  apiRouter.post('/detectors/results/{dataSourceId}/_search/{resultIndex}/{onlyQueryCustomResultIndex}',
```

Comment:
can we be consistent where `dataSourceId` is present in the path? perhaps always earlier/in the beginning, or always at the very end. Suggest to not have in the middle for a few reasons:
- having data source ID before `_search` doesn't make logical sense; the full search path should be the base path before adding optional parameters like datasourceid
- inconsistent, where it is after `results` here, but before `results` on line 97, for example
- more error-prone when constructing the path on client-side to make the calls
- more explicit paths needed. if always at a consistent place in the path, you could have a single utility fn to insert the datasourceid for any given path and greatly reduce all of the explicit ones. As a bonus, if defined under `common/`, it can be used for client and server-side and ensure a lot more consistency

Reply:
I didn't give much thought to where to place dataSourceId in the path since I considered it just a link between the reducer and the server. As long as they aren't conflicting with others, it seemed fine.

I've updated the path to place all dataSourceIds at the end in the latest revision.

Reply:
Cool, sgtm.